Dataset.py

```diff
@@ -152,14 +152,31 @@ def __init__(self, name=None,
       assert sorted(chunk_step.keys()) == sorted(chunk_size.keys())
       assert chunk_step.max_value() > 0, "chunking step must be positive (for some key)"
     self.chunk_step = chunk_step
-    if context_window is None:
-      context_window = NumbersDict(0)
```

Comment:
Do you know why `context_window` was set to `NumbersDict(0)` here and not to `NumbersDict()`? Using the latter resolves the issues in the test cases (in line 164 now).

Reply:
`NumbersDict(0)` and `NumbersDict()` is not the same. We explicitly want that the `broadcast_value` is 0 (i.e. the default fallback value). Just `NumbersDict()` would not have any default fallback value.

Reply:
(Btw, you should have posted the exception somewhere. I don't remember what it was. I assume the problem was maybe related but in fact somewhere else.)

Reply:
======================================================================
FAIL: test_Dataset.test_batches_recurrent_1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 198, in runTest
    line: self.test(*self.arg)
    locals:
      self = <local> test_Dataset.test_batches_recurrent_1
      self.test = <local> <function test_batches_recurrent_1 at 0x7fb689320510>
      self.arg = <local> ()
  File "/work/asr4/vieting/programs/returnn/tests/test_Dataset.py", line 108, in test_batches_recurrent_1
    line: assert_equal(all_batches[0].seqs[0].seq_end_frame, 10)
    locals:
      assert_equal = <global> <bound method TestCase.assertEqual of <nose.tools.trivial.Dummy testMethod=nop>>
      all_batches = <local> [<Batch start_seq:0, len(seqs):1>, <Batch start_seq:0, len(seqs):1>, <Batch start_seq:0, len(seqs):1>, <Batch start_seq:1, len(seqs):1>, <Batch start_seq:1, len(seqs):1>, <Batch start_seq:1, len(seqs):1>], len = 6
      seqs = <not found>
      seq_end_frame = <not found>
  File "/usr/lib/python3.5/unittest/case.py", line 820, in assertEqual
    line: assertion_func(first, second, msg=msg)
    locals:
      assertion_func = <local> <bound method TestCase._baseAssertEqual of <nose.tools.trivial.Dummy testMethod=nop>>
      first = <local> NumbersDict(numbers_dict={'data': 10, 'classes': 10}, broadcast_value=0)
      second = <local> 10
      msg = <local> None
  File "/usr/lib/python3.5/unittest/case.py", line 813, in _baseAssertEqual
    line: raise self.failureException(msg)
    locals:
      self = <local> <nose.tools.trivial.Dummy testMethod=nop>
      self.failureException = <local> <class 'AssertionError'>
      msg = <local> "NumbersDict(numbers_dict={'data': 10, 'classes': 10}, broadcast_value=0) != 10", len = 78
AssertionError: NumbersDict(numbers_dict={'data': 10, 'classes': 10}, broadcast_value=0) != 10

======================================================================
FAIL: test_Dataset.test_batches_non_recurrent_1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 198, in runTest
    line: self.test(*self.arg)
    locals:
      self = <local> test_Dataset.test_batches_non_recurrent_1
      self.test = <local> <function test_batches_non_recurrent_1 at 0x7fb689320598>
      self.arg = <local> ()
  File "/work/asr4/vieting/programs/returnn/tests/test_Dataset.py", line 168, in test_batches_non_recurrent_1
    line: assert_equal(all_batches[0].seqs[0].seq_end_frame, 5)
    locals:
      assert_equal = <global> <bound method TestCase.assertEqual of <nose.tools.trivial.Dummy testMethod=nop>>
      all_batches = <local> [<Batch start_seq:0, len(seqs):1>, <Batch start_seq:0, len(seqs):1>, <Batch start_seq:0, len(seqs):2>, <Batch start_seq:1, len(seqs):1>, <Batch start_seq:1, len(seqs):1>]
      seqs = <not found>
      seq_end_frame = <not found>
  File "/usr/lib/python3.5/unittest/case.py", line 820, in assertEqual
    line: assertion_func(first, second, msg=msg)
    locals:
      assertion_func = <local> <bound method TestCase._baseAssertEqual of <nose.tools.trivial.Dummy testMethod=nop>>
      first = <local> NumbersDict(numbers_dict={'data': 5, 'classes': 5}, broadcast_value=0)
      second = <local> 5
      msg = <local> None
  File "/usr/lib/python3.5/unittest/case.py", line 813, in _baseAssertEqual
    line: raise self.failureException(msg)
    locals:
      self = <local> <nose.tools.trivial.Dummy testMethod=nop>
      self.failureException = <local> <class 'AssertionError'>
      msg = <local> "NumbersDict(numbers_dict={'data': 5, 'classes': 5}, broadcast_value=0) != 5", len = 75
AssertionError: NumbersDict(numbers_dict={'data': 5, 'classes': 5}, broadcast_value=0) != 5

----------------------------------------------------------------------
Ran 9 tests in 0.049s

FAILED (failures=2)

Process finished with exit code 1

Reply:
I think the problem is that since I removed the check `if ctx_lr:` in line 1012, now `start -= self.ctx_left` is executed even if we do not want a context window. 

In that case, the `broadcase_value` of `self.ctx_left` (which is 0) is used and eventually ends up in `all_batches[0].seqs[0].seq_end_frame` which is why the test fails. 

I could include a check again whether `self.ctx_left` and `self.ctx_right` are both zero.

Reply:
(Please properly format using Markdown. I edited your post.)

So, maybe what we want here is actually that `start -= self.ctx_left` would not add a `broadcast_value` to `start`, I guess. I'm not sure if that is always true, or only in this case. Actually, I think calculating `start - self.ctx_left` should result in `NumbersDict(numbers_dict={'data': 5, 'classes': 5}, broadcast_value=0)`, so `NumbersDict.__sub__` (`-`) is correct. However, about the logic for `-=`, I'm not exactly sure... That logic is implemented in `NumbersDict.__isub__`.

Reply:
Btw, just adding such a check is not the right solution. That would be just a workaround.

Reply:
I pushed b8c79207a203c9 (to a test branch albert-numbersdict-better-inplace), let's see if all the tests passes. If so, I will push that to master. This should solve this problem with `NumbersDict`.

Reply:
Ok, this is now failing somewhere else. I have to look into that.

Reply:
This way, `test_batches_non_recurrent_1()` fails because in `EngineBatch.py` -> `add_frames()` -> `self.max_num_frames_per_slice += length`, `NumbersDict(0) += NumbersDict({'data': 5, 'classes': 5})` results in `NumbersDict(0)` which eventually leads to the error. We could replace `a += b` by `a = a + b` there and the tests pass.

However, I am not sure what the most sensible behaviour would be for adding two `NumbersDicts`, one having just the `broadcast_value` and one having just the `dict`. So, should it be `NumbersDict(0) + NumbersDict({'data': 5, 'classes': 5}) = NumbersDict({'data': 5, 'classes': 5}, broadcast_value=0)`? The 0 could be misleading then.

Reply:
Maybe you are right, maybe using `context_window = NumbersDict()` is actually a good solution. But if we use that, for the other code where we set `broadcast_value=0`, we also should remove that.

Reply:
I removed `broadcast_value=0` in the other cases now.